source/dub/project.d

```diff
@@ -1210,6 +1205,91 @@ private string processVars(Project, Package)(string var, in Project project, in
 	} else return var;
 }
 
+/// Expand variables using `$VAR_NAME` or `${VAR_NAME}` syntax.
+/// `$$` escapes itself and is expanded to a single `$`.
+private string expandVars(alias expandVar)(string s)
+{
+	import std.functional : not;
+
+	auto result = appender!string;
+
+	static bool isVarChar(char c)
+	{
+		import std.ascii;
+		return isAlphaNum(c) || c == '_';
+	}
+
+	while (true)
+	{
+		auto pos = s.indexOf('$');
+		if (pos < 0)
+		{
+			result.put(s);
+			return result.data;
+		}
+		result.put(s[0 .. pos]);
+		s = s[pos + 1 .. $];
+		enforce(s.length > 0, "Variable name expected at end of string");
+		switch (s[0])
+		{
+			case '$':
+				result.put("$");
+				s = s[1 .. $];
+				break;
+			case '{':
+				pos = s.indexOf('}');
+				enforce(pos >= 0, "Could not find '}' to match '${'");
+				result.put(expandVar(s[1 .. pos]));
+				s = s[pos + 1 .. $];
+				break;
+			default:
+				pos = s.representation.countUntil!(not!isVarChar);
+				if (pos < 0)
+					pos = s.length;
+				result.put(expandVar(s[0 .. pos]));
+				s = s[pos .. $];
+				break;
+		}
+	}
+}
+
+unittest
+{
+	string[string] vars =
+	[
+		"A" : "a",
+		"B" : "b",
+	];
+
+	string expandVar(string name) { auto p = name in vars; enforce(p, name); return *p; }
+
+	assert(expandVars!expandVar("") == "");
+	assert(expandVars!expandVar("x") == "x");
+	assert(expandVars!expandVar("$$") == "$");
+	assert(expandVars!expandVar("x$$") == "x$");
+	assert(expandVars!expandVar("$$x") == "$x");
+	assert(expandVars!expandVar("$$$$") == "$$");
+	assert(expandVars!expandVar("x$A") == "xa");
+	assert(expandVars!expandVar("x$$A") == "x$A");
+	assert(expandVars!expandVar("$A$B") == "ab");
+	assert(expandVars!expandVar("${A}$B") == "ab");
+	assert(expandVars!expandVar("$A${B}") == "ab");
+	assert(expandVars!expandVar("a${B}") == "ab");
+	assert(expandVars!expandVar("${A}b") == "ab");
+
+	import std.exception : assertThrown;
+	assertThrown(expandVars!expandVar("$"));
+	assertThrown(expandVars!expandVar("${}"));
+	assertThrown(expandVars!expandVar("$|"));
+	assertThrown(expandVars!expandVar("x$"));
+	assertThrown(expandVars!expandVar("$X"));
+	assertThrown(expandVars!expandVar("${"));
+	assertThrown(expandVars!expandVar("${X"));
+
+	// https://github.com/dlang/dmd/pull/9275
+	assert(expandVars!expandVar("$${DUB_EXE:-dub}") == "${DUB_EXE:-dub}");
```

Comment:
Should we test for other Bash variable syntaxes: https://www.tldp.org/LDP/abs/html/parameter-substitution.html ?

Reply:
Well, the point of this test is to explicitly check that Dub doesn't do anything crazy here, and it ignores anything following `$$` and passes it through verbatim, whatever the syntax might be. The previous implementation expected very specific syntax to be used even when escaping Dub processing, which was presumptuous and led to wrong behavior. This test (and any other tests we might add in the same vein) are supplemental / illustrative to the behavior tests above, and don't actually add to the test coverage. An implementation of `expandVars` which somehow passes all of these tests but would behave incorrectly with other kinds of Bash variable substitution syntax would be very weird indeed.

Reply:
Ok, good enough for me.